mne/decoding/time_gen.py

```diff
@@ -226,58 +230,74 @@ def predict(self, epochs):
+            if not np.all([len(test) == len(train) for test in tests]):
+                raise ValueError('`train_times` and `test_times` must '
+                                 'have identical lengths')
 
         # Store all testing times parameters
         self.test_times_ = test_times
 
-        # Prepare parallel predictions across time points
+        n_orig_epochs, _, n_times = X.shape
+
+        # Subselects the to-be-predicted epochs so as to manipulate a
+        # contiguous array X by using slices rather than indices.
+        test_epochs = []
+        if self.predict_mode == 'cross-validation':
+            test_idxs = [ii for train, test in self._cv_splits for ii in test]
+            start = 0
+            for _, test in self._cv_splits:
+                n_test_epochs = len(test)
+                stop = start + n_test_epochs
+                test_epochs.append(slice(start, stop, 1))
+                start += n_test_epochs
+            X = X[test_idxs]
+
+        # Prepare parallel predictions across testing time points
         # FIXME Note that this means that TimeDecoding.predict isn't parallel
-        parallel, p_time_gen, n_jobs = parallel_func(_predict_slices, n_jobs)
-        n_test_slice = max(len(sl) for sl in self.train_times_['slices'])
+        parallel, p_func, n_jobs = parallel_func(_predict_slices, self.n_jobs)
+        n_test_slice = max(len(sl) for sl in self.test_times_['slices'])
         # Loop across estimators (i.e. training times)
         n_chunks = min(n_test_slice, n_jobs)
-        splits = [np.array_split(slices, n_chunks)
+        chunks = [np.array_split(slices, n_chunks)
                   for slices in self.test_times_['slices']]
-        splits = map(list, zip(*splits))
-
-        def chunk_X(X, slices):
-            """Smart chunking to avoid memory overload"""
-            # from object array to list
-            slices = [sl for sl in slices if len(sl)]
-            start = np.min(slices)
-            stop = np.max(slices) + 1
-            slices_ = np.array(slices) - start
-            X_ = X[:, :, start:stop]
-            return (X_, self.estimators_, self.cv_, slices_.tolist(),
-                    self.predict_mode, self.predict_method)
-
-        y_pred = parallel(p_time_gen(*chunk_X(X, slices))
-                          for slices in splits)
-
-        # concatenate chunks across test time dimension. Don't use
-        # np.concatenate as this would need new memory allocations
-        self.y_pred_ = [[test for chunk in train for test in chunk]
-                        for train in map(list, zip(*y_pred))]
+        chunks = map(list, zip(*chunks))
 
-        _warn_once.clear()  # reset self-baked warning tracker
+        # To minimize memory during parallelization, we apply a some chunking
+        y_pred = parallel(p_func(*_chunk_data(
+            X, chunk, self, n_orig_epochs, test_epochs)) for chunk in chunks)
```

Comment:
hmm ... if you unpack the return variables to `_chunk_data` here, it makes one wonder what is passed to `p_func`. For the sake of readability, I would say:

```py
y_pred = parallel(p_func(X_split, gat.estimators_, gat._cv_splits, slices_, gat.predict_mode, gat.predict_method, n_orig_epochs, test_epochs) for X_split in _chunk_data(...)]
```
Aside, why do you need to pass each of the gat attributes separately? Couldn't you simply pass the object?